programs/server/config.xml

```diff
@@ -66,69 +66,289 @@
-         It provides low-level data access between servers.
-         This port should not be accessible from untrusted networks.
-         See also 'interserver_http_credentials'.
-         Data transferred over connections to this port should not go through untrusted networks.
-         See also 'interserver_https_port'.
+
+    <!-- END: Obsolete legacy configurations. Use interface definitions in 'interfaces' instead! -->>
+
+
+
+    <!-- Trusted proxy definitions for specifying in listening interface configs below.
+         See 'allow_proxies' and 'allow_direct' in relevant entries inside 'interfaces' section.
+         Each section can have arbitrary name but names must be unique.
+         The names can be used in 'allow_proxies' section of an interface config.
+         Each section have one mandatory parameter - 'protocol'.
+         Supported protocols are:
+            - PROXY
+            - HTTP
+         You can define any number of proxies for each of the protocols.
       -->
-    <interserver_http_port>9009</interserver_http_port>
+    <proxies>
+
+        <my_load_balancer>
+            <!-- HAProxy PROXY v1/v2 protocol. -->
+            <protocol>PROXY</protocol>
+
+            <!-- The version of the protocol to expect. Can be either 1 or 2.
+                 If not specified, both versions will be allowed and the actual version will be detected automatically.
+              -->
+            <!-- <version>2</version> -->
+
+            <!-- List of IPv4 and/or IPv6 networks, connections from which will be assumed to be from trusted proxies. -->
+            <trust>
+                <net>10.10.10.1/24</net>
+            </trust>
+        </my_load_balancer>
 
-    <!-- Port for communication between replicas with TLS.
-         You have to configure certificate to enable this interface.
-         See the openSSL section below.
-         See also 'interserver_http_credentials'.
+        <my_http_proxy>
+            <!-- Proxy information from X-Forwarded-For HTTP header.-->
+            <protocol>HTTP</protocol>
+
+            <!-- List of IPv4 and/or IPv6 networks, connections from which will be assumed to be from trusted proxies.
+                 Note, that all intermediate peers (but not the initiator client peer) from X-Forwarded-For must be from
+                  some of the networks from this list, for the information about the forwarding chain to be considered authentic.
+              -->
+            <trust>
+                <net>10.10.20.1/24</net>
+            </trust>
+        </my_http_proxy>
+
+    </proxies>
+
+
+    <!-- Interface definitions for ClickHouse server to listen for.
+         Each section can have arbitrary name but names must be unique.
+         The names will be used in logging, as keys for stats, etc.
+         Each section have one mandatory parameter - 'protocol'.
+         Supported protocols are:
+            - native_tcp
+            - native_http
```

Comment:
why native_http? maybe just http?

Reply:
The protocol used here is more than HTTP. It's a API + wire language ClickHouse uses, over HTTP. So this is a ClickHouse native protocol's incarnation over HTTP. Not just faceless vanilla HTTP. (Same applies to TCP.)

Reply:
In most of the places: Native = native protocol of clickhouse run over 9000 port, aka tcp.

http in all places is called http (no need to call it native, it may only lead to confusion). Same for grpc. 
https://clickhouse.tech/docs/en/interfaces/http/

vs 
https://clickhouse.tech/docs/en/interfaces/tcp/

Reply:
The naming you are mentioning is common but inconsistent and confusing. Consider this somewhat new/upgraded naming as another proposed improvement within this PR.

Reply:
Don't see why 'native http' is better than just 'http'. (we don't have any kind of non-native http, or standard http)

Reply:
It's better because HTTP is HTTP, and the protocol ClickHouse is listening over HTTP is some distinct protocol, defined in ClickHouse only. Moreover, there are several other protocols that ClickHouse is handling over HTTP at the same time, like interserver communication, prometheus.

Reply:
> is some distinct protocol, defined in ClickHouse only.

Why do you say so? It's just a regular HTTP. You can use browser / curl / whatever http client.

> several other protocols that ClickHouse is handling over HTTP 

Yes, and they have other names. 

P.S. I am against changing the established terminology and generally accepted names in a single pull request. There must be really important reasons for this. You didn't name any important reasons for now.

Reply:
> Why do you say so? It's just a regular HTTP. You can use browser / curl / whatever http client.

Protocols form a hierarchy (and where they don't, those are places which people often call design failures of the past, compare TCP6 vs TCP4 headers and how they align with IPv6 vs IPv4, you'll see how the clear separation of the concepts opens possibilities for huge optimizations). With the same success you can say, that you can use `netcat` with HTTP interface, since it is also TCP. But that doesn't mean that HTTP interface needs to be renamed to TCP. A lot of problems, both in dealing with as well as maintaining in the code comes from such fallacies and abuses of terminology/abstractions. ClickHouse code suffers from that in many places.

Reply:
> P.S. I am against changing the established terminology and generally accepted names in a single pull request. There must be really important reasons for this. You didn't name any important reasons for now.

https://clickhouse.tech/docs/en/interfaces/tcp/ reads as "Native Interface (TCP)" so there is no substantial change in this PR at all. Just more consistent use.

Reply:
Cluckhouse invented it own (native) protocol which run on the top of tcp. That's why Native TCP.

ClickHouse just implemented Http server (very standard and usual) for those who can't do native

Reply:
Please read the definition of the network protocol, and ask yourself a question: "Does all existing HTTP RFCs fully cover the language ClickHouse speaks on its typical 8123 port?"

Reply:
1) Yes. It follows http standard. 

2) Is the language we speak really English?  :)

My native language is russian. My English is far from perfect, and I'm not following all the grammar rules. Does that mean that I'm talking not English but 'native english'?

Reply:
> ClickHouse uses HTTP as a vehicle

That is correct. Clickhouse http works on the top of normal, standard http. Native protocol for clickhouse is it's own binary, working on the top of http.

> it's more like Simple English

Why not native english? 
Or why not 'simple_http'? :) 

> gRPC is not HTTP/2, is it?..

It works on the the top of http/2. Cluckhouse work on the top of grpc, to implement needed commands/calls. 

> I see you enjoy

I'm wasting time. I see you changing existing naming without good motivation. That may lead to confusions. I will not be able to ask - `do you use clickhouse-native or http protocol' without risks to be misunderstood ('ah but http is also native'). 

That talk goes nowhere. I think we need a third opinion.

Reply:
> It works on the the top of http/2. Clickhouse work on the top of grpc, to implement needed commands/calls.

I was talking about gRPC in general. It works over HTTP/2, but yet it is a separate protocol. Same with ClickHouse over TCP or HTTP or gRPC - all these are separate protocols, and neither is just TCP or HTTP or gRPC.

>  'do you use clickhouse-native or http protocol' 

This is a sloppy and imprecise way of communicating, which is very common in russian-speaking technical community, unfortunately. I want to contribute to fixing this, you want to continue using it because <fill-the-gap>.

> I'm wasting time.

Right, you and I are wasting our both times. I am losing my interest in this functionality, especially in a conditions when I am constantly pushed to do things in a way that I genuinely believe to be wrong. The main arguments I hear are "I don't like it", "it is not how I did things in past."

Reply:
Let's discuss it on the call, or let's just wait for the 3rd opinion.

Reply:
> Про названия - у кх есть протокол придуманный для КХ и используемый для коммуникации между КХ и между КХ клиентом и сервером. Это "главный" / "родной" протокол для КХ. Обычно он на порте 9000 tcp.  Поэтому обычно его называеют "native tcp". Если быть максимально корректным то можно например называть "clickhouse_native". Это протокол 5-7 уровней ISO/OSI на 4м уровне там вовсе не обязательно должен быть tcp (например тот же протокол можно сделать поверх unix socket). 
> 
> Протокол HTTP в КХ - для кликхауса совсем не родной. Именно потому что он следует стандарту HTTP (который не удобен для многих вещей которые  нужны КХ) там многие вещи неудобны. Отклонения от стандарта HTTP там минимальны. На уровне протокола - любой парсер HTTP успешно распарсит весь сетевой обмен КХ по http порту. Из-за мелких отклонений, надстроек - он не перестает быть HTTP. Можно его называть или просто http или clickhouse_http.
> 
> Не надо называть его native_http - в отрыве от контекста КХ это звучит совсем нелепо (у нас что http более native чем у какого-нибудь httpd / nginx?). А в контексте КХ это вызывает путаницу - если http для КХ стал "родным" / "нативным" - то зачем ещё один какой-то 9000 тоже нативный, и почему КХ не пользуется http если это для него нативный. 
> 
> Тоже про grpc. Только там все вооще очень страндартно.

Reply:
I hear your point of view. I don't agree.